ls1intum/Artemis
# `Quiz exercises`: Fix minor issues when participating

### Checklist
#### General



### Description
https://sentry.ase.in.tum.de/organizations/artemis/issues/22481/?project=2 / Closes #5149 
When using the refresh button to refresh to a started quiz things would be in an invalid state after loading the quiz exercise and before loading the quiz participation.

#### Steps for Testing
1. One synchronized Quiz that hasn't started and isn't scheduled
2. Go to the participation page reload it with F5 to break the websocket refresh (see #5024)
3. Start the quiz in another tab
4. use the refresh button in the quiz participation footer
5. See that no errors have occurred in the browser console

#### Manual Tests

### Description
https://sentry.ase.in.tum.de/organizations/artemis/issues/22534/?project=2
When a quiz participation times out without the student ever saving anything, the resulting submission has no submitted answers at all (not even empty ones)
Also the scoring explanation would have an unfilled placeholder for the achieved score.

#### Steps for Testing
1. One batched Quiz (can also happen in individual mode)
2. Add a batch
3. Join it
4. Start batch batch
5. End Now the quiz (or otherwise wait for it to expire without clicking any option or submitting as the student)
6. view the results as the student
7. check that no error happens in the browser console
8. check that the scoring explanation doesn't have an unfilled placeholder

#### Manual Tests

### Description
https://sentry.ase.in.tum.de/organizations/artemis/issues/21981/events/77b3d74488314fef8f49f7b7169e6172/?project=2
This is almost certainly just a user error when using a invalid password for joining batches. I made the error message more specific and also showed it to the user when joining fails.

#### Steps for Testing
1. One batched Quiz (can also happen in individual mode)
2. Fail to join in various ways (wrong/expired password, joined in anther tab, the other tab has already completed the submission)
3. Check the error message (as with submission errors it is not translated, as the error comes from the server)

#### Manual Tests

### Description
https://sentry.ase.in.tum.de/organizations/artemis/issues/20828/?project=2 / Closes #5150
When service workers are disabled update checks would cause an unhandled promise rejection.
When the service worker is disable trigger the update prompt based on the `Content-Version` response header.

#### Steps for Testing
1. It is unclear how exactly service workers get disabled, there does not seem to be an option in any browser to do so
2. At least in firefox unregistering the service worker seems to somewhat disable them for that domain (in chrome the service worker will just come back immediately)
3. In the firefox dev tools click > Application > Service Worker > Unregister to break the service worker
   ![image](https://user-images.githubusercontent.com/44437514/171049247-8ea01229-8c39-47cd-8356-987aaddccb89.png)
4. Navigate artemis a bit / wait
5. check that there are no errors in the console

#### Manual Tests

### Not changed
https://sentry.ase.in.tum.de/organizations/artemis/issues/21981/?project=2
these issues are documented in the code https://github.com/ls1intum/Artemis/blob/4d0d4e5883f927252394e75bdd7533977b1ff216/src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java#L519-L525
and fixing that is a significantly bigger job that deserves its own PR

### Review Progress





#### Code Review
